### ISSUE TITLE
questions on reliable_endpoint_t structure

### ISSUE THREAD

**Issue by mreinstein** [networkprotocol/reliable.io#2]
while looking at:

```c
struct reliable_endpoint_t
{
   .
   .
   .
    uint16_t * acks;
    uint16_t sequence;
    struct reliable_sequence_buffer_t * sent_packets;
    struct reliable_sequence_buffer_t * received_packets;
```

I don't understand the usage `sequence` and `sent_packets` as used on these lines:

https://github.com/networkprotocol/reliable.io/blob/master/reliable.c#L725
https://github.com/networkprotocol/reliable.io/blob/master/reliable.c#L733

Is this redundant? Why effectively store 3 sequence numbers when the 2 above lines seem to conflate `sequence` with `sent_packets->sequence` ?

**Comment by gafferongames** [comment]
Thanks! I'll take a look today

**Comment by gafferongames** [comment]
Hi, there is no redundancy here:

1. endpoint::sequence is the current send sequence #

2. endpoint::sent_packets is a sliding window containing data per-sent packet, indexed by packet sequence #

3. endpoint::received_packets is a sliding window containing data per-received packet, indexed by sequence number (which corresponds to the endpoint->sequence number on the other side, *not* the endpoint->seqence on the receiver side).

cheers

**Comment by mreinstein** [comment]
thanks for the info 👍

**Comment by gafferongames** [comment]
No worries. The key is to understand the endpoint acts as both the sender and receiver simultaneously. It's a bit hard to parse because of that. -- cheers

**Comment by mreinstein** [comment]
Hey Glenn, congrats on getting funding for your new startup! You must be
excited!

-Mike

**Comment by gafferongames** [comment]
Yep! Thanks!
